### PR TITLE
Host mapping analysis and design

### PR DESCRIPTION
**Pull request by NajmudheenCT**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:

 /kind design
 /kind documentation

**What this PR does / why we need it**:
Bring out analysis output and design for unified host mapping model in delfin
**Which issue(s) this PR fixes**:

Fixes #

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:

**Special notes for your reviewer**:

### CODE REVIEW

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+| initiator     | yes      | yes      | yes      | yes     | yes     | yes                                                             |
+| port          | yes      | no       | no       | no      | no      | no                                                              |
+| port_group    | yes      | no       | no       | no      | no      | no                                                              |
+| storage_group | yes      | optional | no       | no      | no      | optional ( vnx storage group is a collection of hosts and luns) |
+| volume        | yes      | yes      | yes      | yes     | yes     | yes                                                             |
+
+## Proposed Delfin common model 
+
+
+![](./Resources/Delfin_host_mapping.png)
```
Comment (kumarashit) [inline specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md:92]: Few questions on the Model:
If a Vendor doesn't support StorageGroup or PortGroup or HostGroup, then will we create a dummy group for Maksing View?
Can we segregate properly the masking and mapping views and then connect?
Keep the first class objects as the entities directly involved in making or mapping. Like: MaskingView will contain 1) Port 2) HBA 3) LUN 
Now all the groups like HostGroup, StorageGroup and PortGroup will have second level relationship.

One catch here is that, MaskingView can get bloated if we keep LUN ID there. To solve it maybe we can keep only the ArrayPort and HBA in the Masking view. Volume/LUN info table will have PortID associated. So all info can be retrieved though SQL Join.
HostGroup/PortGroup/StorageGroup will be under the HBA/ArrayPort/LUN.

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+
+### HDS VSP
+Each initiator is considers as host, mutiple HBA is same host may be represented as mutiple hosts. host name is optional. Host Group ( Initiator group) is supported
+![](./Resources/vsp_host_mapping.png)
+
+## Delfin host mapping objects
+### Masking view
+The masking view identifies the host group (set of hosts), the port group (set of array ports), and the storage group (set of volumes) in a combination which are associated. 
+### Host group
+logical grouping of hosts
```
[the remark this answers is not in the document]
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md:68]: goup of hosts, host will have mutiple HBAs, host group will have mutiple hosts

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+### Port group
+Logical grouping of array ports .
```
[the remark this answers is not in the document]
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md:70]: yes

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md (hunk 1 above)
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md:97]: I will add some explanation with work flows to give more clarity

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+## Host Mapping Models of different Storage devices
+
+
+
+
+
+### DELL EMC VMAX 
+VMAX uses masking view to represnt the Host to LUN access mechanism
+
+![](./Resources/vmax_host_mapping.png)
```
[the remark this answers is not in the document]
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: done

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+
+### DELL EMC VNX 
+
+![](./Resources/vnx_host_mapping.png)
```
[the remark this answers is not in the document]
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: On VNX when you add LUN to a storage group you essentially map that LUN to all ports in the array (FC and iSCSI). Now when you add host(s) to a storage group you perform masking, you allow specific set of WWNs to see LUNs that are part of that storage group.

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md (hunk 1 above)
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: added more ino with workflow..

Comment (kumarashit) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: Do we really need: StorageGroupVolumeMapping, MaskingViewStorageGroupMapping, MaskingViewPortGroupMapping,  HostGroupHostMapping, MaskingViewHostGroupMapping ? Will the DB query be slow due to many joins?

Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: we should have a way to represent one to many relation, rhese mapping table represent the relation. Volume can be part of mutiple storage groups, simlary host to host group and port to port gorup

File: specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+### Host
+Represents a host , consists of mutiple initiators
+### Initiator
+An HBA wwn or iscsi traget.
+### Port
+Array port through the volume is mapped to the initiator
+### Volume
+volume or lun or logical device or vitual volume which is mapped to initiators.
+
+## Host mapping object requirement mapping to some storage models
```
[the remark this answers is not in the document]
Comment (NajmudheenCT) [reply specs/SIM/Delfin_Host_Mapping_Model_Analysis_and_Design.md]: changed to "Delfin host mapping models to storage backend availability matrix " , does it give more clarity ?

**Comment by NajmudheenCT** [comment]
Closing this PR to raise new design PR based on template.

**Pull request closed without merging**